File: src/main/python/com/hazza/leetcode_by_python/SegmentTwo/No67_Add_Binary.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/17 19:59
# https://leetcode.com/problems/add-binary/


class Solution:
    def addBinary(self, a: str, b: str) -> str:
        if len(a) == 0:
            return b
        if len(b) == 0:
            return a

        if a[-1] == '1' and b[-1] == '1':
            return self.addBinary(self.addBinary(a[:-1], b[:-1]), '1') + '0'
        elif a[-1] == '0' and b[-1] == '0':
            return self.addBinary(a[:-1], b[:-1]) + '0'
        else:
            return self.addBinary(a[:-1], b[:-1]) + '1'


if __name__ == '__main__':
    solution = Solution()

    assert solution.addBinary('11', '1') == '100'
    assert solution.addBinary('1010', '1011') == '10101'
    assert solution.addBinary('1', '111') == '1000'

File: src/main/java/com/hazza/leetcode/SegmentTwo/No67AddBinary.java
package com.hazza.leetcode.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/add-binary/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-21
 * Time: 10:43 AM
 */
public class No67AddBinary {
    public String addBinary(String a, String b) {
        StringBuilder res = new StringBuilder();
        int i = a.length() - 1, j = b.length() - 1;
        int flag = 0;
        while (i >= 0 || j >= 0) {
            int sum = flag;
            if (i >= 0) sum += a.charAt(i--) - '0';
            if (j >= 0) sum += b.charAt(j--) - '0';
            res.append(sum % 2);
            flag = sum / 2;
        }
        if (flag == 1) res.append(flag);

        return res.reverse().toString();
    }
}
